db/records/operations/insert.py

```diff
@@ -98,6 +100,26 @@ def insert_from_select(from_table, target_table, engine, col_mappings=None):
         ins = target_table.insert().from_select(target_table_col_list, sel)
         try:
             result = conn.execute(ins)
+        except IntegrityError as e:
+            if type(e.orig) == NotNullViolation:
+                raise NotNullError
+            elif type(e.orig) == ForeignKeyViolation:
+                raise ForeignKeyError
+            elif type(e.orig) == UniqueViolation:
+                # ToDo: Try to differentiate between the types of unique violations
+                # Scenario 1: Adding a duplicate value into a column with uniqueness constraint in the target table.
+                # Scenario 2: Adding a non existing value twice in a column with uniqueness constraint in the target table.
+                # Both the scenarios currently result in the same exception being thrown.
+                raise UniqueValueError
+            elif type(e.orig) == ExclusionViolation:
+                raise ExclusionError
+            else:
+                raise e
+        except ProgrammingError as e:
+            if type(e.orig) == DatatypeMismatch:
+                raise TypeMismatchError
+            else:
+                raise e
         except Exception as e:
             # ToDo raise specific exceptions
             raise e
```

Comment:
I think this is redundant, correct? I.e., if you don't catch the exception, it will be raised anyway.

Reply:
Yes, That is true, But can we be sure that all the unhandled exceptions will get caught in `IntegrityError` and `ProgrammingError`?

Reply:
My point is that this line checks, and then _raises_ the exception if it's unhandled. It doesn't actually handle it. If this `except` block were missing, the behavior of the function would be precisely to raise the exception `e` under these circumstances.

Reply:
Oh, I see your point, refactoring it.